mapproxy/service/ogcapi/map.py

```diff
@@ -0,0 +1,619 @@
+        if layer.is_opaque(query):
+            actual_layers = OrderedDict()
+        for layer_name, map_layers in layer.map_layers_for_query(query):
+            actual_layers[layer_name] = map_layers
+
+    render_layers = []
+    for layers in actual_layers.values():
+        render_layers.extend(layers)
+
+    raise_source_errors = True if server.on_error == "raise" else False
+    renderer = LayerRenderer(
+        render_layers,
+        query,
+        req,
+        raise_source_errors=raise_source_errors,
+        concurrent_rendering=server.concurrent_layer_renderer,
+    )
+
+    merger = LayerMerger()
+    renderer.render(merger)
+
+    if server.attribution and server.attribution.get("text"):
+        merger.add(attribution_image(server.attribution["text"], query.size))
+
+    try:
+        result = merger.merge(
+            size=query.size, image_opts=image_opts, bbox=query.bbox, bbox_srs=query.srs
+        )
+    except Exception as ex:
+        raise OGCAPIServer.exception(
+            "InternalError",
+            "error while merging layers: %s" % str(ex),
+            status=500,
+        )
+
+    if query != orig_query:
+        result = SubImageSource(
+            result, size=orig_query.size, offset=offset, image_opts=image_opts
+        )
+
+    # Provide the wrapping WSGI app or filter the opportunity to process the
+    # image before it's wrapped up in a response
+    result = server.decorate_img(
+        result,
+        "wms.map",
+        list(actual_layers.keys()),
+        req.environ,
+        (query.srs.srs_code, query.bbox),
+    )
+
+    query = orig_query
+
+    try:
+        result.georef = GeoReference(bbox=query.bbox, srs=query.srs)
+        result_buf = result.as_buffer(image_opts)
+    except IOError as ex:
+        raise OGCAPIServer.exception(
+            "InternalError",
+            "error while processing image file: %s" % ex,
+            status=500,
+        )
+
+    return Response(
+        result_buf,
+        content_type=FORMAT_TYPES[query.format],
+        headers=get_headers(server, query),
+        status=200,
+    )
+
+
+def get_map(
+    server: OGCAPIServer, req: Request, coll_id: Optional[str], map_filename: str
+):
+    log = server.log
+    log.debug(f"Map for {coll_id}")
+
+    allowed_args = set(
+        [
+            "f",
+            "crs",
+            "bbox-crs",
+            "bbox",
+            "center-crs",
+            "center",
+            "subset",
+            "subset-crs",
+            "width",
+            "height",
+            "scale-denominator",
+            "mm-per-pixel",
+            "bgcolor",
+            "transparent",
+        ]
+    )
+    collections = req.args.get("collections", None)
+
+    if coll_id is None and (server.default_dataset_layers or collections):
+        allowed_args.add("collections")
+
+    if req.args.get("service", None) == "WMS":
+        # Hack for the OpenLayers map in our collection.html that issues WMS requests...
```

Comment:
We will need a fixup for this in a follow up PR, just adding comment as a reminder for myself